class.jetpack-xmlrpc-server.php

```diff
@@ -177,52 +181,117 @@ public function remote_provision( $request ) {
+		);
+	}
+
+	/**
+	 * This XML-RPC method is called from the /jpphp/provision endpoint on WPCOM in order to
+	 * register this site so that a plan can be provisioned.
+	 *
+	 * @param array $request An array containing at minimum a nonce key and a local_username key.
+	 *
+	 * @return WP_Error|array
+	 */
+	public function remote_provision( $request ) {
+		$user = $this->fetch_and_verify_local_user( $request );
 
 		if ( ! $user ) {
-			return $this->error( new Jetpack_Error( 'user_unknown', 'User not found.', 404 ) );
+			return $this->error( new WP_Error( 'input_error', __( 'Valid user is required', 'jetpack' ) ), 'jpc_remote_register_fail' );
 		}
 
-		require_once JETPACK__PLUGIN_DIR . '_inc/class.jetpack-provision.php';
+		$site_icon = ( function_exists( 'has_site_icon' ) && has_site_icon() )
+			? get_site_icon_url()
+			: false;
 
-		wp_set_current_user( $user->ID );
+		$auto_enable_sso = ( ! Jetpack::is_active() || Jetpack::is_module_active( 'sso' ) );
 
-		// Filter allowed parameters.
-		$allowed_provision_args = array(
-			'wpcom_user_id',
-			'wpcom_user_email',
-			'local_username',
-			'plan',
-			'force_register',
-			'force_connect',
-			'onboarding',
-			'partner_tracking_id',
-		);
+		/** This filter is documented in class.jetpack-cli.php */
+		if ( apply_filters( 'jetpack_start_enable_sso', $auto_enable_sso ) ) {
+			$redirect_uri = add_query_arg(
+				array(
+					'action'      => 'jetpack-sso',
+					'redirect_to' => rawurlencode( admin_url() ),
+				),
+				wp_login_url() // TODO: come back to Jetpack dashboard?
+			);
+		} else {
+			$redirect_uri = admin_url();
+		}
 
-		$args = array_intersect_key(
-			$request,
-			array_flip( $allowed_provision_args )
+		// generate secrets
+		$role    = Jetpack::translate_user_to_role( $user );
+		$secrets = Jetpack::init()->generate_secrets( 'authorize', $user->ID );
+
+		$response = array(
+			'jp_version'   => JETPACK__VERSION,
+			'redirect_uri' => $redirect_uri,
+			'user_id'      => $user->ID,
+			'user_email'   => $user->user_email,
+			'user_login'   => $user->user_login,
+			'scope'        => Jetpack::sign_role( $role, $user->ID ),
+			'secret'       => $secrets['secret_1'],
+			'is_active'    => Jetpack::is_active(),
 		);
 
-		$result = Jetpack_Provision::register_and_build_request_body( $args );
+		if ( $site_icon ) {
+			$response['site_icon'] = $site_icon;
+		}
 
-		if ( is_wp_error( $result ) ) {
-			return $this->error( $result, 'jpc_remote_provision_fail' );
+		// this feels out of place here now
+		if ( $request['onboarding'] ) {
+			Jetpack::create_onboarding_token();
+			$response['onboarding_token'] = Jetpack_Options::get_option( 'onboarding' );
+		}
+
+		return $response;
+	}
+
+	private function fetch_and_verify_local_user( $request ) {
+		if ( empty( $request['local_user'] ) ) {
+			return $this->error(
+				new Jetpack_Error(
+					'local_user_missing',
+					esc_html__( 'The required "local_user" parameter is missing.', 'jetpack' ),
```

Comment:
Let's use `__()` here so that the quotes don't end up as `&quot;` in the AIP response.